crates/wasmtime/src/module.rs

```diff
@@ -371,6 +335,43 @@ impl Module {
         Ok(())
     }
 
+    /// Ahead-of-time (AOT) compiles a WebAssembly module.
+    ///
+    /// The `bytes` provided must be in one of two formats:
+    ///
+    /// * A [binary-encoded][binary] WebAssembly module. This is always supported.
+    /// * A [text-encoded][text] instance of the WebAssembly text format.
+    ///   This is only supported when the `wat` feature of this crate is enabled.
+    ///   If this is supplied then the text format will be parsed before validation.
+    ///   Note that the `wat` feature is enabled by default.
+    ///
+    /// See [`Module::new`] for errors that may be returned by this function.
+    ///
+    /// [binary]: https://webassembly.github.io/spec/core/binary/index.html
+    /// [text]: https://webassembly.github.io/spec/core/text/index.html
+    pub fn compile(engine: &Engine, bytes: &[u8], mut output: impl Write) -> Result<()> {
+        const USE_PAGED_MEM_INIT: bool = cfg!(all(feature = "uffd", target_os = "linux"));
+
+        if bytes.starts_with(COMPILED_MODULE_HEADER) {
+            bail!("input is already a compiled module");
+        }
+
+        #[cfg(feature = "wat")]
+        let bytes = wat::parse_bytes(&bytes)?;
+
+        let (_, artifacts, types) =
+            CompilationArtifacts::build(engine.compiler(), &bytes, USE_PAGED_MEM_INIT)?;
+
+        // Write a header that marks this as a compiled module
+        output.write_all(COMPILED_MODULE_HEADER)?;
+        bincode_options().serialize_into(
+            output,
+            &SerializedModule::from_artifacts(engine.compiler(), &artifacts, &types),
+        )?;
```

Comment:
Could this method perhaps be written as `Module::new(engine, bytes)?.serialize_into(output)`?

I'd ideally prefer to avoid having this method since it seems like it should be a trivial combination of a constructor plus a `serialize` invocation, and splitting it up shouldn't be all that much less ergonomic. (I'd imagine it's fine to add the compiled module header to serialized modules too and we could even support them in `Module::new` to automatically `deserialize` as well)

Reply:
Also I've personally never been really that sold on the `impl Write` output for this sort of use case. It's hard for me to imagine where the output is so large it must be streamed rather than being held resident in memory.

Reply:
Ah I see, those are good points. I think though I'd ideally still like to push on a small set of constructors for `Module` instead of a standalone `compile` function. Could we perhaps defer the registration of JIT'd code for the platform to instantiation? I'm imagining we could have a `Mutex` of some kind which mutates the state of `Module` the first time it's instantiated or something like that.

Personally I think it'd be a bummer if, when using cross-compilation, you're forced to never touch `Module::new` and friends and you instead have to use `Module::compile`, but there's no real fundamental reason to do so per se.

Another possible alternative is that if the module looks cross-compiled we just never do the registration bits (and somehow we should error-out in `Instantiate` if the target actually mismatches the current host...), but I think I'd prefer to have the defer for now to avoid registering jit code on hosts where you don't actually do any instantiation.

Reply:
I'm still inclined to believe that constructing a `Module` implies it's usable (i.e.instantiatable) for the current host.  I think the constructors should error if 1) the module is AOT-compiled for a different target or 2) the associated `Config` has a target other than host and thus JIT doesn't make sense.

I'm also inclined to believe that it's not desirable to produce a `Module` for AOT compilation (even internally) as the point of AOT compilation is getting an artifact that can be loaded _at a later time_ and likely from a different host.  I think that anyone making use of `Module::compile` would not be using `Module::new` on the artifact from the same host as doing AOT compilation and loading it on the same host isn't gaining anything over the code cache.

It's not just code registration that we're skipping in `Module::compile` currently, but all the work done to get the module ready to be executed (i.e. "linking it" - copying the text sections into executable memory, applying relocations, etc.), none of which is required for AOT compilation.

To me, at least, AOT compilation is a pretty explicit action that's independent of constructing a `Module` that can be executed.  Perhaps `Module::compile` isn't the best place for this, but I am strongly in favor of now allowing a `Module` to construct if the target isn't for host.

Reply:
My main worry is that having `Module::{new, compile}` could possibly be confusing for users because they both look like constructors and most might even lean towards "compile" since wasm is generally understood to be compiled to native code. Perhaps though we can solve all this by tweaking things a bit API-wise.

In a hypothetical future version of the `wasmtime` crate you'll be able to jettison the runtime entirely and instead only have the cranelift compiler (and vice versa, you could only have the runtime without the compiler). In this hypothetical world if `Module` is a "ready to instantiate module" then we wouldn't want to expose it from the version of the crate that is missing the runtime. That would put `Module::compile` in a weird place since it needs to exist but the type it's attached to doesn't want to exist.

Perhaps we could solve this by moving the method to `Engine::compile`? That way it's theoretically far away enough api-wise to not confuse users, we'll still have an `Engine` even with all modes of the wasmtime crate (whether you're removing the runtime or not), and we can clearly document that it's only intended for ahead-of-time-compiled cases.

I think though that I would still want to tweak things a bit ideally:

* I think the format of the data from `Module::compile` and `Module::serialize` should be the same
* We could probably remove `Module::deserialize` and just document more sources of binary data for `Module::new` and/or `from_binary`
* Personally I would still like to remove the `impl Write` in favor of just returning a `Vec<u8>`. I think that working with a `Vec<u8>` is generally more convenient and if we really need it for performance we could always add something like `compile_into` or `serialize_into`.

How does that all sound?

Reply:
I'm on-board with everything you just proposed; I agree that moving `compile` out of `Module` into (perhaps) `Engine::compile_module` would make it clear that this is not useful for constructing a `Module` itself.

I also am not opposed to using `Vec<u8>` instead of `impl Write`; agree that it's easy enough to enhance if necessary in the future.

Reply:
Oh actually a nice name might be `precompile_module` or something like that to emphasize it a bit more?

Reply:
Even better!